EIPS/eip-4524.md

```diff
@@ -0,0 +1,82 @@
+---
+eip: 4524
+title: Safer ERC20
+description: Extending ERC20 with ERC165 and adding safeTransfer (like ERC721 and ERC1155)
+author: William Schwab (@wschwab)
+discussions-to: https://ethereum-magicians.org/t/why-isnt-there-an-erc-for-safetransfer-for-erc20/7604
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-12-05
+requires: 20, 165
+---
+
+## Abstract
+
+This standard extends [ERC20](./erc-20.md) tokens with [EIP-165](./eip-165.md), and adds familiar functions from [ERC-721](./eip-721.md) and [EIP-1155](./eip-1155.md) ensuring receiving contracts have implemented proper functionality.
+
+## Motivation
+
+[EIP-165](./eip-165.md) adds (among other things) the ability to tell if a target recipient explicitly signals compatibility with an ERC. This is already used in the core EIPs for NFTs, [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md). In addition, EIP-165 is a valuable building block for extensions on popular standards to signal implementation, a trend we've seen in a number of NFT extensions. This EIP aims to bring these innovations back to ERC20.
+
+The importance of [EIP-165](./eip-165.md) is perhaps felt most for app developers looking to integrate with a generic standard such as EIP-20 or EIP-721, while integrating newer innovations built atop these standards. An easy example would be token permits, which allow for a one-transaction approval and transfer. This has already been implemented in many popular ERC20 tokens using the [EIP-2612](./eip-2612.md) standard or similar. A platform integrating ERC20 tokens has no easy way of telling if a particular token has implemented token permits or not. (As of this writing, EIP-2612 does not require EIP-165.) With EIP-165, the app (or contracts) could query `supportsInterface` to see if the `interfaceId` of a particular EIP is registered (in this case, EIP-2612), allowing for easier and more modular functions interacting with ERC20 contracts. It is already common in NFT extensions to include an EIP-165 interface with a standard, we would argue this is at least in part due to the underlying [ERC-721](./eip-721.md) and [EIP-1155](./eip-1155.md) standards integrating EIP-165. Our hope is that this extension to ERC20 would also help future extensions by making them easier to integrate.
+
+## Specification
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
```

Comment:
```suggestion
The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
```

Reply:
Per RFC 2119, it is "key words" https://www.ietf.org/rfc/rfc2119.txt

Reply:
there's a history here of splitting it into two words, though I see your point